### ISSUE TITLE
12$: warn on missing semicolon before (expression)

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#1090]
```javascript
    nextConfig.push({
      proxied_site: "http://localhost:1337",
      port: 8084,
      early_hints: [
        {
          type: "style",
          url: "/style.css",
        },
        {
          type: "script",
          url: "/script.js",
        },
      ],
    })  // WARNING: Missing semicolon or confusing indentation
    ({ statusLine, headers } = await makeProxyRequest());
```
```javascript

foo()  // WARNING: Missing semicolon or confusing indentation
[a, b].forEach(f)
```

**Comment by strager** [comment]
Idea for the logic for when to report the warning:
1. Found a `(` or `[` in `parse_expression_remainder`
2. `(` or `[` has a leading newline
3. The `(`'s or `[`'s column is equal to the statement start's column

Logic 3 is hard to implement currently because quick-lint-js does not track column numbers. Perhaps we can get away with just logic 1 and 2. We would need to check the rate of false positives to decide whether we need logic 3 or not.

**Comment by strager** [comment]
We need to make sure that a semicolon is legal before reporting the warning. In the following example, a semicolon is not legal before `(y)` or `[z]`, so we should not report any diagnostic:
```javascript
let x = (
  f()
  (y)
  [z]
);
```

**Comment by samir-hamud** [comment]
Hi, is there a deadline? 
if you don't mind me taking more time to deliver the task, I'll take it
